core/src/main/java/org/sonarsource/sonarlint/core/BindingClueProvider.java

```diff
@@ -0,0 +1,254 @@
+import org.sonarsource.sonarlint.core.repository.connection.SonarQubeConnectionConfiguration;
+
+import static java.util.stream.Collectors.toSet;
+import static org.apache.commons.lang.StringUtils.removeEnd;
+import static org.apache.commons.lang.StringUtils.trimToNull;
+import static org.sonarsource.sonarlint.core.commons.log.SonarLintLogger.singlePlural;
+import static org.sonarsource.sonarlint.core.repository.connection.SonarCloudConnectionConfiguration.SONARCLOUD_URL;
+
+public class BindingClueProvider {
+
+  private static final SonarLintLogger LOG = SonarLintLogger.get();
+  private static final String SONAR_SCANNER_CONFIG_FILENAME = "sonar-project.properties";
+  private static final String AUTOSCAN_CONFIG_FILENAME = ".sonarcloud.properties";
+
+  private final ConnectionConfigurationRepository connectionRepository;
+  private final SonarLintClient client;
+
+  public BindingClueProvider(ConnectionConfigurationRepository connectionRepository, SonarLintClient client) {
+    this.connectionRepository = connectionRepository;
+    this.client = client;
+  }
+
+  public List<BindingClueWithConnections> collectBindingCluesWithConnections(String configScopeId, Set<String> connectionIds) throws InterruptedException {
+    var bindingClues = collectBindingClues(configScopeId);
+    return matchConnections(bindingClues, connectionIds);
+  }
+
+  private List<BindingClueWithConnections> matchConnections(List<BindingClue> bindingClues, Set<String> eligibleConnectionIds) {
+    LOG.debug("Match connections...");
+    List<BindingClueWithConnections> cluesAndConnections = new ArrayList<>();
+    for (BindingClue bindingClue : bindingClues) {
+      var connectionsIds = matchConnections(bindingClue, eligibleConnectionIds);
+      if (!connectionsIds.isEmpty()) {
+        cluesAndConnections.add(new BindingClueWithConnections(bindingClue, connectionsIds));
+      }
+    }
+    LOG.debug("{} {} having at least one matching connection", cluesAndConnections.size(), singlePlural(cluesAndConnections.size(), "clue", "clues"));
+    return cluesAndConnections;
+  }
+
+  public static class BindingClueWithConnections {
+    private final BindingClue bindingClue;
+    private final Set<String> connectionIds;
+
+    private BindingClueWithConnections(BindingClue bindingClue, Set<String> connectionIds) {
+      this.bindingClue = bindingClue;
+      this.connectionIds = connectionIds;
+    }
+
+    public BindingClue getBindingClue() {
+      return bindingClue;
+    }
+
+    public Set<String> getConnectionIds() {
+      return connectionIds;
+    }
+  }
+
+  private List<BindingClue> collectBindingClues(String checkedConfigScopeId) throws InterruptedException {
+    LOG.debug("Query client for binding clues...");
+    FindFileByNamesInScopeResponse response;
+    try {
+      response = client.findFileByNamesInScope(new FindFileByNamesInScopeParams(checkedConfigScopeId, List.of(SONAR_SCANNER_CONFIG_FILENAME, AUTOSCAN_CONFIG_FILENAME))).get(1,
```

Comment:
This will block the event bus thread for max 1min, not sure it's a problem or not (or maybe not now)

Reply:
Good catch, I will move that to a separate thread.